java-checks/src/main/java/org/sonar/java/checks/tests/TooManyAssertionsCheck.java

```diff
@@ -0,0 +1,123 @@
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks.tests;
+
+import java.util.HashMap;
+import java.util.Map;
+import org.sonar.check.Rule;
+import org.sonar.check.RuleProperty;
+import org.sonar.java.model.ModifiersUtils;
+import org.sonar.plugins.java.api.JavaFileScanner;
+import org.sonar.plugins.java.api.JavaFileScannerContext;
+import org.sonar.plugins.java.api.semantic.Symbol;
+import org.sonar.plugins.java.api.tree.BaseTreeVisitor;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.MethodReferenceTree;
+import org.sonar.plugins.java.api.tree.MethodTree;
+import org.sonar.plugins.java.api.tree.Modifier;
+import org.sonar.plugins.java.api.tree.Tree;
+
+import static org.sonar.java.checks.helpers.UnitTestUtils.ASSERTION_INVOCATION_MATCHERS;
+import static org.sonar.java.checks.helpers.UnitTestUtils.ASSERTION_METHODS_PATTERN;
+import static org.sonar.java.checks.helpers.UnitTestUtils.REACTIVE_X_TEST_METHODS;
+import static org.sonar.java.checks.helpers.UnitTestUtils.TEST_METHODS_PATTERN;
+import static org.sonar.java.checks.helpers.UnitTestUtils.isUnitTest;
+import static org.sonar.java.model.ExpressionUtils.methodName;
+
+@Rule(key = "S5961")
+public class TooManyAssertionsCheck extends BaseTreeVisitor implements JavaFileScanner {
+
+  private static final int DEFAULT_MAX = 10;
+
+  @RuleProperty(key = "MaximumAssertionNumber", description = "The maximum authorized number of assertions in a test method", defaultValue = "" + DEFAULT_MAX)
+  public int maximumAssertionNumber = DEFAULT_MAX;
+
+  private final Map<Symbol, Integer> assertionsInMethod = new HashMap<>();
+  private JavaFileScannerContext context;
+
+  @Override
+  public void scanFile(JavaFileScannerContext context) {
+    this.context = context;
+    assertionsInMethod.clear();
+    scan(context.getTree());
+    assertionsInMethod.clear();
+  }
+
+  @Override
+  public void visitMethod(MethodTree methodTree) {
+    if (ModifiersUtils.hasModifier(methodTree.modifiers(), Modifier.ABSTRACT)) {
+      return;
+    }
+
+    if (isUnitTest(methodTree) && numberOfAssertions(methodTree.symbol()) > maximumAssertionNumber) {
+      context.reportIssue(this, methodTree.simpleName(), String.format("Refactor this method in order to have less than %d assertions.", maximumAssertionNumber));
```

Comment:
2 comments:
* With an `IssuableSubscriptionVisitor`, you could directly call `reportIssue(methodTree.simpleName())`

* I think it would be valuable to use secondaries on all the assertions, in order to proof our count.